crates/wasi-nn/src/backend/winml.rs

```diff
@@ -0,0 +1,165 @@
+    ) -> Result<Graph, BackendError> {
+        let model = read(&path.join("mobilenet.onnx"))?;
+        self.load(&[&model], target)
+    }
+}
+
+struct WinMLGraph {
+    model: LearningModel,
+    device_kind: LearningModelDeviceKind,
+}
+
+unsafe impl Send for WinMLGraph {}
+unsafe impl Sync for WinMLGraph {}
+
+impl BackendGraph for WinMLGraph {
+    fn init_execution_context(&self) -> Result<ExecutionContext, BackendError> {
+        let device = LearningModelDevice::Create(self.device_kind.clone())?;
+        let session = LearningModelSession::CreateFromModelOnDevice(&self.model, &device)?;
+        let box_: Box<dyn BackendExecutionContext> = Box::new(WinMLExecutionContext::new(session));
+        Ok(box_.into())
+    }
+}
+
+struct WinMLExecutionContext {
+    session: LearningModelSession,
+    binding: LearningModelBinding,
+    result: Option<LearningModelEvaluationResult>,
+}
+
+impl WinMLExecutionContext {
+    fn new(session: LearningModelSession) -> Self {
+        Self {
+            binding: LearningModelBinding::CreateFromSession(&session).unwrap(),
+            session,
+            result: None,
+        }
+    }
+}
+
+impl BackendExecutionContext for WinMLExecutionContext {
+    fn set_input(&mut self, index: u32, tensor: &Tensor) -> Result<(), BackendError> {
+        let input = self.session.Model()?.InputFeatures()?.GetAt(index)?;
+        // TODO: Bind input.
+        unsafe {
+            let data = std::slice::from_raw_parts(
+                tensor.data.as_ptr() as *const f32,
+                tensor.data.len() / 4,
+            );
+
+            self.binding.Bind(
+                &input.Name()?,
+                &TensorFloat::CreateFromArray(
+                    &input.cast::<TensorFeatureDescriptor>()?.Shape()?,
+                    data,
+                )?,
+            )?;
+        }
+        Ok(())
+    }
+
+    fn compute(&mut self) -> Result<(), BackendError> {
+        self.result = Some(self.session.Evaluate(&self.binding, &HSTRING::new())?);
+        Ok(())
+    }
+
+    fn get_output(&mut self, index: u32, destination: &mut [u8]) -> Result<u32, BackendError> {
+        if self.result.is_none() {
+            return Err(BackendError::BackendAccess(anyhow::Error::msg(
+                "Output is not ready.",
+            )));
+        }
+        let output_name = self.session.Model()?.OutputFeatures()?.GetAt(index)?;
+        let output_name_hstring = output_name.Name()?;
+
+        // Print results.
+        let vector_view = self
+            .result
+            .as_ref()
+            .unwrap()
+            .Outputs()?
+            .Lookup(&output_name_hstring)?
+            .cast::<TensorFloat>()?
+            .GetAsVectorView()?;
+        let output: Vec<f32> = vector_view.into_iter().collect();
+        unsafe {
+            destination.copy_from_slice(std::slice::from_raw_parts(
+                output.as_ptr() as *const u8,
+                output.len() * 4,
+            ));
```

Comment:
I believe this is another place where the tensor item type is hard-coded: we need to be able to handle other types.

Reply:
4 is the size of TensorFloat, which is the output data type for probability.